File: frontend/types/bot.ts
/**
 * Type definitions for bot-related entities.
 */

export enum BotType {
  TELEGRAM_USERBOT = "telegram_userbot",
  TELEGRAM_BOT = "telegram_bot",
  DISCORD_BOT = "discord_bot",
}

export enum BotStatus {
  STOPPED = "stopped",
  STARTING = "starting",
  RUNNING = "running",
  STOPPING = "stopping",
  CRASHED = "crashed",
}

export interface Bot {
  id: string;
  name: string;
  type: BotType;
  config: Record<string, any>;
  status: BotStatus;
  auto_restart: boolean;
  created_at: string;
  updated_at: string;
  last_started_at: string | null;
  process_id: number | null;
  restart_count: number;
  last_crash_at: string | null;
}

export interface BotCreate {
  name: string;
  type: BotType;
  config: Record<string, any>;
  auto_restart: boolean;
}

export interface BotUpdate {
  name?: string;
  config?: Record<string, any>;
  auto_restart?: boolean;
}

export interface BotListResponse {
  total: number;
  page: number;
  page_size: number;
  bots: Bot[];
}

export interface BotStatusResponse {
  id: string;
  name: string;
  status: BotStatus;
  process_id: number | null;
  uptime_seconds: number | null;
  last_started_at: string | null;
}

File: backend/app/schemas/bot.py
"""Pydantic schemas for bot operations."""

from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict

from app.models.bot import BotType, BotStatus


class BotCreate(BaseModel):
    """Schema for creating a new bot."""
    name: str = Field(..., min_length=1, max_length=100, description="Unique bot name")
    type: BotType = Field(..., description="Bot type (telegram_userbot, telegram_bot, discord_bot)")
    config: Dict[str, Any] = Field(..., description="Bot configuration (tokens, settings)")
    auto_restart: bool = Field(True, description="Enable automatic restart on crash")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "My Telegram Bot",
                "type": "telegram_bot",
                "config": {
                    "token": "123456:ABC-DEF1234ghIkl-zyx57W2v1u123ew11",
                    "admin_user_ids": [12345678]
                },
                "auto_restart": True
            }
        }
    )


class BotUpdate(BaseModel):
    """Schema for updating bot configuration."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    config: Optional[Dict[str, Any]] = None
    auto_restart: Optional[bool] = None

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "Updated Bot Name",
                "auto_restart": False
            }
        }
    )


class BotResponse(BaseModel):
    """Schema for bot response."""
    id: str
    name: str
    type: BotType
    config: Dict[str, Any]
    status: BotStatus
    auto_restart: bool
    created_at: datetime
    updated_at: datetime
    last_started_at: Optional[datetime]
    process_id: Optional[int]
    restart_count: int
    last_crash_at: Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class BotListResponse(BaseModel):
    """Schema for paginated bot list."""
    total: int
    page: int
    page_size: int
    bots: List[BotResponse]


class BotStatusResponse(BaseModel):
    """Schema for bot status information."""
    id: str
    name: str
    status: BotStatus
    process_id: Optional[int]
    uptime_seconds: Optional[int]
    last_started_at: Optional[datetime]
